src/apps/service/apps/de/metadata.clj

```diff
@@ -153,19 +156,36 @@
             :value (config/workspace-metadata-beta-attr-label)
             :unit  "attr"}]})
 
+(defn- community-name->admin-set
+  [username community-name]
+  (->> (communities/get-community-admin-set username community-name)
+       (map #(vector community-name %))))
+
+(defn- notify-community-admins
+  [username app-name community-names]
+  (let [admin->communities-map (->> community-names
+                                    (mapcat (partial community-name->admin-set username))
+                                    (group-by second))]
+    (doseq [[admin community+admin-lists] admin->communities-map]
+      (notifications/send-community-admin-notification app-name
+                                                       (groups-client/lookup-subject username admin)
+                                                       (map first community+admin-lists)))))
```

Comment:
These functions need to transform a list of community names, each with a list of their respective admins, into a list of those unique admins, each with their communities. Then the admins can be notified that the app was added to the communities they administrate.

So given the list of community names extracted from the app's published metadata, first the admins of each community is fetched from the `iplant-groups` service.

For example, given communities `com1`, `com2`, and `com3`, the admins may be `user1` and `user2` for `com1`, `user2` and `user3` for `com2`, and `user3` and `user4` for `com3`. As the admins are fetched for each community, they're combined into a list of community-admin lists, such as `[["com1" "user1"], ["com1" "user2"]]`, then combined into 1 list by `mapcat`:
```clojure
[["com1" "user1"], ["com1" "user2"], ["com2" "user2"], ["com2" "user3"], ["com3" "user3"], ["com3" "user4"]]
```

Then `group-by` turns this into a map keyed by each unique admin:
```clojure
{"user1" [["com1" "user1"]], "user2" [["com1" "user2"] ["com2" "user2"]], "user3" [["com2" "user3"] ["com3" "user3"]], "user4" [["com3" "user4"]]}
```

Finally, `doseq` iterates over each admin-list/key-value pair, passing the admin and the community list (extracted with `map first`) to the notifications function.

I'm open to suggestions if there's a better way to do this 🤔 
Otherwise, I can add some of this explanation to these function docs.

Reply:
A Clojure bike shedding puzzle! This sounded like fun! I played around with this for a while, and found another way to do it, but I haven't decided whether or not this was a better way to do it yet.

Stubbed-out functions:

```clojure
(def admins-for {"com1" #{"user1" "user2"} "com2" #{"user2" "user3"} "com3" #{"user3" "user4"}})

(defn get-community-admin-set [_ community-name]
  (admins-for community-name #{}))

(defn send-community-admin-notification [app-name admin community-names]
  (println admin ", " app-name "was added to the following communities:" (clojure.string/join ", " community-names)))
```
Implementation functions (with namespace aliases removed for testing):

```clojure
(defn- community-name->communities-by-admin [username community-name]
  (into {} (map #(vector % [community-name])
                (get-community-admin-set username community-name))))

(defn- group-communities-by-admin [username community-names]
  (apply merge-with into (map (partial community-name->communities-by-admin username) community-names)))

(defn- notify-community-admins [username app-name community-names]
  (doseq [[admin communities-for-admin] (group-communities-by-admin username community-names)]
    (send-community-admin-notification app-name admin communities-for-admin)))
```
Testing:

```
user=> (group-communities-by-admin "sarahr" ["com1" "com2"])
{"user2" ["com1" "com2"], "user1" ["com1"], "user3" ["com2"]}
user=> (notify-community-admins "sarahr" "app1" ["com1" "com2"])
user2 ,  app1 was added to the following communities: com1, com2
user1 ,  app1 was added to the following communities: com1
user3 ,  app1 was added to the following communities: com2
nil
user=> (notify-community-admins "sarahr" "app1" ["com1" "com2" "com3"])
user2 ,  app1 was added to the following communities: com1, com2
user1 ,  app1 was added to the following communities: com1
user3 ,  app1 was added to the following communities: com2, com3
user4 ,  app1 was added to the following communities: com3
```

It seems to work and I think that it's fairly easy to read. The number of functions is possibly a bit higher than optimal, but I always tend to err on the side of having too many functions rather than too few. (I'm still deciding whether or not that's a good thing. 😆)

Reply:
Here's a slight variation on my suggestion from earlier. Again, I'll leave thoughts about the relative merits of this approach to others. The same stubbed-out functions as before are assumed to be defined:

```clojure
(defn- notify-community-admins [username app-name community-names]
  (->> (map (partial community-name->communities-by-admin username) community-names)
       (apply merge-with into)
       (map (partial apply send-community-admin-notification app-name))
       dorun))
```

Reply:
okay, not entirely sure if this is better now that I've written it out, but, replacing only @slr71's `group-communities-by-admin` (and removing the need for `community-name->communities-by-admin`):

```clojure
(defn group-communities-by-admin [username communities]
  (reduce 
    (fn [final-community-map community]
        (merge-with into
          final-community-map
          (reduce
            (fn [admin-map admin] (assoc admin-map admin (conj (get admin-map admin []) community)))
            {}
            (get-community-admin-set username community))))
    {}
    communities))
```

Pretty similar, really :laughing: other than using `reduce` more. Anyway, maybe an interesting permutation.

Reply:
I'm getting the feeling that whatever approach floats your boat most, some good comments would probably be a good idea, anyway. Seems like all the ideas use enough clojure magic to benefit.

Reply:
Heh, here's a slight variation on Ian's idea:

```clojure
(defn group-communities-by-admin [username communities]
  (reduce
   (fn [community-map community]
     (reduce (fn [community-map admin] (update community-map admin #(conj (vec %) community)))
             community-map
             (get-community-admin-set username community)))
   {}
   communities))
```

Reply:
Nested reduction is always mind bending for me, so I thought I should play with it a little to force myself to wrap my brain around the concept.

Reply:
Here's what I've been playing with:

```clojure
(defn- notify [username app-name community-names]
  (->> community-names
        (mapcat (fn [community] (map #(hash-map % [community]) (get-community-admin-set username community))))
        (apply merge-with into)
        (map (partial apply send-community-admin-notification app-name))
        dorun)))
```

building off @slr71's idea using threading, which I think can make it nice. Not super fond of the map-inside-mapcat thing though, so, hm.

Reply:
And just for good measure, the macroexpanded-and-lightly-edited version of the above:

```clojure
(dorun
 (map
  (partial apply send-community-admin-notification app-name)
  (apply
   merge-with
   into
   (mapcat
    (fn
     [community]
     (map
      (fn [admin] {admin [community]})
      (get-community-admin-set username community)))
    community-names))))
```

Reply:
Here's a slight variation on that using `zipmap` that I kind of like, although it suffers from the map-inside-map problem as well:

```clojure
(defn- notify-community-admins [username app-name community-names]
  (->> (map (fn [community-name]
              (zipmap (get-community-admin-set username community-name)
                      (repeat [community-name])))
            community-names)
       (apply merge-with into)
       (map (partial apply send-community-admin-notification app-name))
       dorun))
```

Reply:
Heh, this version _looks_ cleaner, but I'm not sure that it's easier to understand:

```clojure
(defn- notify-community-admins [username app-name community-names]
  (->> (map (juxt (partial get-community-admin-set username) (comp repeat vector)))
       (map (partial apply zipmap))
       (apply merge-with into)
       (map (partial apply send-community-admin-notification app-name))
       dorun))
```

Reply:
> Here's a slight variation on that using `zipmap` that I kind of like, although it suffers from the map-inside-map problem as well:
> 
> ```clojure
> (defn- notify-community-admins [username app-name community-names]
>   (->> (map (fn [community-name]
>               (zipmap (get-community-admin-set username community-name)
>                       (repeat [community-name])))
>             community-names)
>        (apply merge-with into)
>        (map (partial apply send-community-admin-notification app-name))
>        dorun))
> ```

Thanks @slr71 ! I agree with @ianmcorvidae (as mentioned in slack) that this is the most readable, and I think it's the cleanest suggestion, so I'll go with the variation that you 2 came up with in chat:
```clojure
(defn map-keys-to-value [set key]
  (zipmap set (repeat key)))

(defn- notify-community-admins [username app-name community-names]
  (->> (map #(map-keys-to-value (get-community-admin-set username %) [%]) community-names)
       (apply merge-with into)
       (map (partial apply send-community-admin-notification app-name))
       dorun))
```

Though now that we're not creating an initial `{"community-name" #{admins}}` map anymore, I should rename `map-keys-to-value` 🤔